### PR TITLE
feat: channels with messages linked to user

### PR DESCRIPTION
**Pull request by FlynnOConnell**
Addition of channels. 

- Each channel stores a group of messages
- Each message is tied to a user
- If a user is deleted, the message is still stored, with the user_id set to "deleted"

### CODE REVIEW

File: backend/migrations/20230804194015_add_user_to_message.up.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+-- Add up migration script here
+ALTER TABLE message
+ADD COLUMN user_id BIGINT NOT NULL,
+ADD COLUMN channel_id BIGINT NOT NULL,
+ADD CONSTRAINT fk_user FOREIGN KEY (user_id) REFERENCES fluke_user(id),
+ADD CONSTRAINT fk_channel FOREIGN KEY (channel_id) REFERENCES channel(id);
```
Comment (LukeBrandon) [inline backend/migrations/20230804194015_add_user_to_message.up.sql]:
```suggestion
ADD COLUMN user_id NOT NULL FOREIGN KEY REFERENCES fluke_user(id),
ADD COLUMN channel_id NOT NULL FOREIGN KEY REFERENCES channel(id);
```
Think you can just do this.

File: backend/migrations/20230804194015_add_user_to_message.down.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+-- Add down migration script here
+ALTER TABLE message
+DROP CONSTRAINT IF EXISTS fk_user,
+DROP CONSTRAINT IF EXISTS fk_channel,
```
Comment (LukeBrandon) [inline backend/migrations/20230804194015_add_user_to_message.down.sql]:
```suggestion
```
If you put constraints on the value directly you can get rid of this

File: backend/src/models/message.rs [hunk 3]
```diff
@@ -9,6 +9,7 @@ pub struct CreateMessageSchema {
 #[derive(Debug, Deserialize, Serialize, FromRow)]
 pub struct MessageModel {
     pub id: i64,
+    pub user_id: i64,
```
Comment (LukeBrandon) [inline backend/src/models/message.rs]: you need to update the message create to include the channel message

File: backend/migrations/20230804194015_add_user_to_message.down.sql (hunk 2 above)
Comment (LukeBrandon) [reply backend/migrations/20230804194015_add_user_to_message.down.sql]: The constraint applies to the column, should just be able to drop column?

File: backend/migrations/20230804194015_add_user_to_message.up.sql (hunk 1 above)
Comment (LukeBrandon) [inline backend/migrations/20230804194015_add_user_to_message.up.sql]:
```suggestion
ADD CONSTRAINT FOREIGN KEY (user_id) REFERENCES fluke_user(id),
ADD CONSTRAINT FOREIGN KEY (channel_id) REFERENCES channel(id);
```
you dont need to have these names do you?

File: backend/src/models/user.rs [hunk 4]
```diff
@@ -30,6 +31,7 @@ pub struct UserModel {
     pub last_name: String,
     pub email: String,
     pub password: String,
+    pub deleted: Option<bool>,
```
Comment (LukeBrandon) [inline backend/src/models/user.rs]:
```suggestion
    pub deleted: bool,
```

File: backend/src/main.rs [hunk 5]
```diff
@@ -64,6 +64,14 @@ async fn main() {
         .route("/users/:id", get(controllers::user::get_user))
         .route("/users/:id", put(controllers::user::update_user))
         .route("/users/:id", delete(controllers::user::delete_user))
+        .route("/channels", get(controllers::channel::list_channels))
+        .route("/channels", post(controllers::channel::create_channel))
+        .route("/channels/:id", get(controllers::channel::get_channel))
+        .route("/channels/:id", put(controllers::channel::update_channel))
+        .route(
+            "/channel/:id",
```
Comment (LukeBrandon) [inline backend/src/main.rs]:
```suggestion
            "/channels/:id",
```

File: backend/src/controllers/user.rs [hunk 6]
```diff
@@ -134,7 +136,14 @@ pub async fn delete_user(
         .await
         .map_err(|_| CustomError::UserNotFound(id.to_string()))?;
 
-    Ok((StatusCode::OK, Json(json!({"message": "User deleted"}))))
+    let _ = sqlx::query!(
+        "UPDATE message SET user_id = NULL WHERE user_id = $1",
+        id
+    )
+    .execute(&pool)
+    .await;
```
Comment (LukeBrandon) [inline backend/src/controllers/user.rs]: This won't work, `user_id` isn't nullable.  I would just do a soft delete on the user model and update them to be `deleted=true` so you don't need to mess with cascading for now.  Then your `list` will have to filter out the deleted ones.

File: backend/src/controllers/channel.rs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    (StatusCode::OK, Json(task))
+}
+
+pub async fn get_channel(
+    Path(id): Path<i64>,
+    Extension(pool): Extension<PgPool>,
+) -> Result<Json<ChannelModel>, CustomError> {
+    let sql = "SELECT * FROM channel where id = ($1)";
+
+    let channel: ChannelModel = sqlx::query_as::<_, ChannelModel>(sql)
```
Comment (LukeBrandon) [inline backend/src/controllers/channel.rs:31]: use `sqlx::query_as!` macro

File: backend/src/controllers/user.rs (hunk 6 above)
Comment (LukeBrandon) [inline backend/src/controllers/user.rs]:
```suggestion
```
Actually you already added a deleted, I'd just remove this then

File: backend/migrations/20230804194015_add_user_to_message.down.sql (hunk 2 above)
Comment (LukeBrandon) [reply backend/migrations/20230804194015_add_user_to_message.down.sql]: you dont need to drop the constraints if you're dropping the column

**Review by LukeBrandon — CHANGES_REQUESTED** [review]
few things

File: backend/src/controllers/message.rs [hunk 8]
```diff
@@ -65,7 +65,7 @@ pub async fn update_message(
 ) -> Result<(StatusCode, Json<MessageModel>), CustomError> {
     let updated = sqlx::query_as!(
         MessageModel,
-        "UPDATE message SET message=$1 WHERE id=$2 RETURNING *",
+        "UPDATE message SET message=$1 WHERE id=$2 RETURNING id, message, created_at, user_id, channel_id",
```
Comment (LukeBrandon) [inline backend/src/controllers/message.rs]:
```suggestion
        "UPDATE message SET message=$1 WHERE id=$2 RETURNING *",
```

File: backend/src/controllers/message.rs [hunk 9]
```diff
@@ -89,8 +89,8 @@ pub async fn create_message(
 
     let created = sqlx::query_as!(
         MessageModel,
-        "INSERT INTO message (message, user_id, channel_id) VALUES ($1, $2, $3) RETURNING *",
-        &message.message, &message.user_id, &message.channel_id
+        "INSERT INTO message (message)  VALUES ($1) RETURNING *",
+        &message.message
```
Comment (LukeBrandon) [inline backend/src/controllers/message.rs]: dis broken

File: backend/src/controllers/user.rs [hunk 10]
```diff
@@ -121,29 +121,12 @@ pub async fn delete_user(
...
-    )
-    .execute(&pool)
-    .await;
+    let _ = sqlx::query_as!(UserModel,"UPDATE fluke_user SET deleted=true WHERE id=($1)", id)
+    .fetch_one(&pool)
+    .await
+    .map_err(|_| CustomError::UserNotFound(id.to_string()))?;
 
-    Ok((StatusCode::OK, Json(json!({"message": "User deleted", "user_id": user_id_str}))))
+    Ok((StatusCode::OK, Json(json!({"message": "User soft-deleted", "user_id": id.to_string()}))))
```
Comment (LukeBrandon) [inline backend/src/controllers/user.rs]:
```suggestion
    Ok((StatusCode::OK, Json(json!({"message": "User deleted", "user_id": id.to_string()}))))
```
They don't need to know its soft

File: backend/src/controllers/message.rs (hunk 9 above)
Comment (LukeBrandon) [reply backend/src/controllers/message.rs]: you need to add `user_id` and `channel_id` here

**Review by FlynnOConnell — COMMENTED** [review]
Fixed multiple issues

**Review by LukeBrandon — CHANGES_REQUESTED** [review]
Gotta run `cargo sqlx prepare`

**Review by LukeBrandon — APPROVED** [review]

**Pull request merged**